Use `interleave()` in CSSSelector

nt1m opened the pull request:
#### 01bcff3058440abda06beecdd2df3b7e3c3930f1
<pre>
Use `interleave()` in CSSSelector
<a href="https://bugs.webkit.org/show_bug.cgi?id=278268">https://bugs.webkit.org/show_bug.cgi?id=278268</a>
<a href="https://rdar.apple.com/134101084">rdar://134101084</a>

Reviewed by Darin Adler.

More efficient to build the string.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):

Canonical link: <a href="https://commits.webkit.org/282390@main">https://commits.webkit.org/282390@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7751253a034d1d47e8f32023ea281aaf6c939dfc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63058 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42414 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15654 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67079 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13662 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65178 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50101 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13946 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50807 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9419 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66127 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/39391 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54597 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31493 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36076 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11944 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12538 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57616 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12273 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68774 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7004 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11886 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58122 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7036 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54669 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58330 "Found 2 new API test failures: /WebKitGTK/TestContextMenu:/webkit/WebKitWebView/popup-event-signal, /WebKitGTK/TestContextMenu:/webkit/WebKitWebView/context-menu-key (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13983 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5838 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38234 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39314 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40425 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39056 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -428,21 +428,21 @@ static void outputNthChildAnPlusB(const CSSSelector& selector, StringBuilder& bu
-    StringBuilder builder;
-
-    auto serializeIdentifierOrStar = [&] (const AtomString& identifier) {
+    auto serializeIdentifierOrStar = [&](const AtomString& identifier, StringBuilder& builder) {
```
Maybe we can write `[]` instead of `[&]` now since this doesn’t need to capture anything?

nt1m replied:
Do I not still need it to capture `serializeIdentifier` ?

nt1m replied:
Actually I don't.

darinadler commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -536,28 +536,20 @@ String CSSSelector::selectorText(StringView separator, StringView rightSide) con
             case PseudoElement::ViewTransitionImagePair:
             case PseudoElement::ViewTransitionOld:
             case PseudoElement::ViewTransitionNew: {
```
No need for braces any more since we don’t have local variables. Same is true elsewhere above and below in this function.

darinadler commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -536,28 +536,20 @@ String CSSSelector::selectorText(StringView separator, StringView rightSide) con
+                ')');
                 break;
             }
             case PseudoElement::Part: {
```
No need for braces any more since we don’t have local variables.

nt1m committed:
Use `interleave()` in CSSSelector
https://bugs.webkit.org/show_bug.cgi?id=278268
rdar://134101084

Reviewed by Darin Adler.

More efficient to build the string.

* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):

Canonical link: https://commits.webkit.org/282390@main
